core/server/common/src/main/java/alluxio/master/metastore/BlockStore.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.metastore;
+
+import alluxio.master.metastore.BlockStore.Block;
+import alluxio.proto.meta.Block.BlockLocation;
+import alluxio.proto.meta.Block.BlockMeta;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * The block store keeps track of block sizes and block locations.
+ *
+ * BlockStores require external synchronization for operations on the same block id. Operations on
+ * different block ids can be performed concurrently.
+ */
+@NotThreadSafe
+public interface BlockStore extends Iterable<Block> {
+  /**
+   * @param id a block id
+   * @return the block's metadata, or empty if the block does not exist
+   */
+  Optional<BlockMeta> getBlock(long id);
+
+  /**
+   * Adds block metadata to the block store. If the block already exists, its metadata will be
+   * updated to the given metadata.
+   *
+   * @param id the block id
+   * @param meta the block metadata
+   */
+  void putBlock(long id, BlockMeta meta);
+
+  /**
+   * Removes a block, or does nothing if the block does not exist.
+   *
+   * @param id a block id to remove
+   */
+  void removeBlock(long id);
+
+  /**
+   * Removes all metadata from the block store.
+   */
+  void clear();
+
+  /**
+   * Gets locations for a block. If the block does not exist or has no locations, an empty list is
+   * returned.
```

Comment:
Will we ever need to distinguish between the two cases?

Reply:
We distinguish today by checking `getBlock(id).isPresent()`

Reply:
What's the locking requirement for the `BlockStore`, do we need to obtain a lock on the block id before doing any operation, or only on multiple operations on the same id?
If we need to call `getBlock` and `getLocations` together, we would need to lock before calling `getLocations` making this call the odd one out.
Breaking the abstraction for a moment, is this because some implementations of `getLocations` cannot distinguish between no locations and no block?

Reply:
The interface-level javadoc tries to lay this out:
```java
/**
 * The block store keeps track of block sizes and block locations.
 *
 * BlockStores require external synchronization for operations on the same block id. Operations on
 * different block ids can be performed concurrently.
 */
@NotThreadSafe
public interface BlockStore extends Iterable<Block> {
```

BlockStore marks itself as `@NotThreadSafe` to communicate that it doesn't take responsibility for thread safety.

> If we need to call getBlock and getLocations together, we would need to lock before calling getLocations making this call the odd one out.
To distinguish between block not existing and block existing, we would need to hold a lock across the calls to `getBlock` and `getLocations`. It's the same for any other operation where you want to check whether the block exists and then do something based on the block existing or not existing.

We're making a tradeoff here where we don't perform the checks needed to allow us to fail fast in cases of metadata inconsistency due to bugs in our code. This is because requiring the sanity checks could be expensive, like in the case where the metadata is split up across a KV store.